### PR TITLE
Add validations for client_id and redirect_uri

### PR DESCRIPTION
**Pull request by callahad**

### CODE REVIEW

**Comment by callahad** [comment]
@djc if you have a moment to review this, I'd greatly appreciate it. :)

File: src/validation.rs [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+impl From<url::ParseError> for ValidationError {
+    fn from(err: url::ParseError) -> ValidationError {
+        ValidationError::Parse(err)
+    }
+}
+
+// -- Validation Functions --
+
+/// Test that a URI is valid and conforms to our expetations.
+pub fn valid_uri(raw_uri: &str) -> Option<ValidationError> {
```
Comment (djc) [inline src/validation.rs]: `Option<ValidationError>` doesn't feel like an idiomatic return type for this, I think `Result<(), ValidationError>` would be clearer even though it's a bit more typing.

File: src/lib.rs [hunk 2]
```diff
@@ -39,7 +39,9 @@ pub mod email;
 pub mod oidc;
 pub mod store;
 pub mod store_cache;
+#[macro_use]
 pub mod validation;
+use validation::{valid_uri, only_origin, same_origin};
```
Comment (djc) [inline src/lib.rs]: I'd squash this commit into the previous one.

File: src/validation.rs (hunk 1 above)
Comment (callahad) [inline src/validation.rs]: `Ok(())` felt weird, but you're probably right about being idiomatic. Rebasing, squashing, and adjusting to `Result<(), ValidationError>`

**Comment by callahad** [comment]
Ready for re-review

**Comment by callahad** [comment]
Whoops, didn't update the unit tests from `is_some` to `is_err`. Fixed, and actually ready for re-review now :)

**Pull request merged**